### PR TITLE
Add deprecation notes for developers.

### PR DESCRIPTION
**Pull request by wlynch**
# Changes

Add deprecation notes for developers.

These are not meant to be hard rules, but general tips / things to keep in mind for deprecations.

/kind documentation

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: api_compatibility_policy.md [hunk 1]
```diff
@@ -151,3 +151,68 @@ Tekton Pipelines may introduce breaking changes to its Go client libraries, as l
...
+
+3. Tombstone API fields before removal.
+
+   Removing fields from the API is most disruptive since it affects both the
+   server and clients, but **clients are not guaranteed to run at the same
+   version as the server**. This can break an integration's ability to safely
+   talk to older API servers.
+
+   A question to ask yourself is: "if I remove this field, can this client still
+   be used with the oldest supported LTS server?". If the answer is no, removing
```
Comment (JeromeJu) [inline api_compatibility_policy.md:207]: Thanks @wlynch 

Shall we suggest there to be a hard earliest timeline that the fields could be removed after defaulting the feature flag in one LTS? And I have been wondering if this also needs to be updated in the compatibility policy?

**Review by lbernick — APPROVED** [review]
thanks! this is great

**Review by JeromeJu — APPROVED** [review]

File: api_compatibility_policy.md (hunk 1 above)
Comment (wlynch) [reply api_compatibility_policy.md:207]: I think this depends on how the field was deprecated.

Most conservative is wait until the last LTS where a user could downgrade cycles out (worst case 1 yr from release). At minimum I'd probably consider it a separate deprecation event from the feature flag removal.

There are probably things you could do to reduce this. For example for the embedded status deprecation - what made this more painful was that the field **might** not be there. If we guaranteed earlier that the childRefs would always be present (since it would have been an additive backwards compatible change) and scoped the deprecation around the opt-out flag we probably could have removed the feature flag and the API field at the same time since we could guarantee childrefs to always be present in older LTS versions.

File: api_compatibility_policy.md [hunk 2]
```diff
@@ -151,3 +151,68 @@ Tekton Pipelines may introduce breaking changes to its Go client libraries, as l
 do not impact users' yaml/json CRD definitions. For example, a change that renames a field of a CRD
 would need to follow the policy on [backwards incompatible changes](#backwards-incompatible-changes),
 but a change that renames the Go struct type for that field is allowed.
+
+## Notes for Developers
```
[the remark this answers is not in the document]
Comment (wlynch) [reply api_compatibility_policy.md:155]: Done.

File: docs/developers/deprecations.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+   own deprecation notification.
+
+   A conservative deprecation sequence to move from field A -> B might look
+   like:
+
+   1. (inital state) Return field A
+   2. Introduce new field B with opt-in feature flag (disabled by default).
+   3. Set feature flag to return both A and B in responses by default.
+   4. Set feature flag to opt-in users by default and only return field B.
+   5. Remove feature flag.
```
Comment (jerop) [inline docs/developers/deprecations.md:32]: to clarify, are all these steps within the 9 months deprecation period? just want to be clear that we won't double the deprecation time

**Review by jerop — COMMENTED** [review]
Thank you @wlynch for clarifying the deprecation process!

I had clarifying questions, mainly wondering if the alpha compatibility policy should apply equally for opt-in and opt-out features (now that features will become opt-out earlier with more users using the feature by default)

**Comment by jerop** [comment]
cc @tektoncd/core-maintainers

File: docs/developers/deprecations.md (hunk 3 above)
Comment (wlynch) [reply docs/developers/deprecations.md:32]: There's no one right answer here - this is going to depend on what the change is, how the deprecation is done, and how disruptive it is.

You can make the argument that both the change in default behavior of a feature flag and the removal of a feature flag should be separate deprecation events. This is especially true for any deprecation that results in xor behavior.

[I think there is room for shortcuts](https://github.com/tektoncd/pipeline/pull/6208#discussion_r1114719628), so long as we have reasonable confidence that clients will be able to notice and adapt to changes - but I think how this can be done will differ on a case by case basis.

**Comment by wlynch** [comment]
> if the alpha compatibility policy should apply equally for opt-in and opt-out features

I think Kubernetes has a reasonable policy for this that we may be able to borrow - https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages

tl;dr 

alpha features == opt in
beta features == opt out
GA features == no feature flag

**Comment by jerop** [comment]
> 
> alpha features == opt in
> beta features == opt out
> GA features == no feature flag
> 

@wlynch I like that K8s policy, let's borrow it! What do other @tektoncd/core-maintainers think? 

Also can we make this a tekton-wide policy (in tektoncd/community), instead of just for tektoncd/pipeline?

**Comment by wlynch** [comment]
> Also can we make this a tekton-wide policy (in tektoncd/community), instead of just for tektoncd/pipeline?

This is probably going to be a little tricky since I don't think there is a consistent definition of alpha/beta/GA across projects (which is by design - not every project wants to adopt the same deprecation policy as pipelines), but you're welcome to!

Let's lead by example here first - we can add to community later.

**Comment by jerop** [comment]
Thought that definitions of alpha/beta/GA in [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md) applied across all projects, but sounds good to me to learn from adoption in this project first

File: docs/developers/deprecations.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+   ```go
+   // Foo is a ...
+   //
+   // Deprecated: use Bar instead
+   type Foo struct {...}
+   ```
+
+   See https://github.com/golang/go/wiki/Deprecated for more details.
+
+3. Tombstone API fields before removal.
```
Comment (dibyom) [inline docs/developers/deprecations.md:48]: 💯

**Review by dibyom — APPROVED** [review]
